Use new JIT helpers for instantiating and continuation thunks

MichalStrehovsky opened the pull request:
Takes advantage of the new functionality added in #121799.

Cc @dotnet/ilc-contrib @jakobbotsch

MichalStrehovsky committed:
Use new JIT helpers for instantiating thunks

MichalStrehovsky committed:
Use new JIT helpers for continuations

jakobbotsch commented:
Can you change this code to `assert(node->AsCall()->IsAsync());`:
https://github.com/dotnet/runtime/blob/bf17d45e7223638964621d28f62b4eb995286560/src/coreclr/jit/lower.cpp#L5881-L5887

Also the part in the comment above about NAOT can be removed.

MichalStrehovsky committed:
Refactor async call handling in lower.cpp

Removed comments about marking calls as async for NativeAOT and added an assertion to ensure the call is async.
